Plausible/DeriveArbitrary.lean

```diff
@@ -0,0 +1,281 @@
+
+```lean
+#eval Arbitrary.runArbitrary (α := Tree) 10
+```
+
+For debugging purposes, we also provide a command elaborator for the `#derive_arbitrary` command:
+
+```lean
+-- `#derive_arbitrary` derives an instance of `Arbitrary` for `Tree`
+#derive_arbitrary Tree
+```
+
+Running this command prints the code for the derived generator to stdout.
+(Note that the definition of the derived generator is already registered at this point.)
+
+## Main definitions
+* Deriving handler for `ArbitrarySized` typeclass
+* `#derive_arbitrary` command elaborator
+
+-/
+
+namespace Plausible
+
+open Arbitrary Idents
+
+/-- Takes the name of a constructor for an algebraic data type and returns an array
+    containing `(argument_name, argument_type)` pairs.
+
+    If the algebraic data type is defined using anonymous constructor argument syntax, i.e.
+    ```
+    inductive T where
+      C1 : τ1 → … → τn
+      …
+    ```
+    Lean produces macro scopes when we try to access the names for the constructor args.
+    In this case, we remove the macro scopes so that the name is user-accessible.
+    (This will result in constructor argument names being non-unique in the array
+    that is returned -- it is the caller's responsibility to produce fresh names,
+    e.g. using `Idents.genFreshNames`.)
+-/
+def getCtorArgsNamesAndTypes (ctorName : Name) : MetaM (Array (Name × Expr)) := do
+  let ctorInfo ← getConstInfoCtor ctorName
+
+  forallTelescopeReducing ctorInfo.type fun args _ => do
+    let mut argNamesAndTypes := #[]
+    for arg in args.toList do
+      let localDecl ← arg.fvarId!.getDecl
+      let mut argName := localDecl.userName
+      -- Check if the name has macro scopes
+      -- If so, remove them so that we can produce a user-accessible name
+      -- (where macro scopes don't appear in the name)
+      if argName.hasMacroScopes then
+        argName := Name.eraseMacroScopes argName
+      argNamesAndTypes := Array.push argNamesAndTypes (argName, localDecl.type)
+
+    return argNamesAndTypes
+
+/-- Produces an instance of the `ArbitrarySized` typeclass for an inductive type
+    whose name is given by `targetTypeName`.
+
+    (Note: the main logic for determining the structure of the derived generator
+    is contained in this function.) -/
+def mkArbitrarySizedInstance (targetTypeName : Name) : CommandElabM (TSyntax `command) := do
+  -- Obtain Lean's `InductiveVal` data structure, which contains metadata about
+  -- the type corresponding to `targetTypeName`
+  let inductiveVal ← getConstInfoInduct targetTypeName
+
+  -- Fetch the ambient local context, which we need to produce user-accessible fresh names
+  let localCtx ← liftTermElabM $ getLCtx
+
+  -- Produce a fresh name for the `size` argument for the lambda
+  -- at the end of the generator function, as well as the `aux_arb` inner helper function
+  let freshSizeIdent := mkIdent $ localCtx.getUnusedName `size
+  let freshSize' := mkIdent $ localCtx.getUnusedName `size'
```

Comment:
The standard way to create unused names in macros is to use `Core.mkFreshUserName`.